### DISCUSSION TITLE
请问如何让 GL M3 只回答固定领域内的问题？ 我这个提示词似乎完全不工作

### DISCUSSION THREAD

**Discussion by chopin1998** [zai-org/ChatGLM3#516] [General]
![image](https://github.com/THUDM/ChatGLM3/assets/643792/8de83386-db0c-4073-bbdf-d06d0217c143)

[
    {
        "content": "你的名字叫 郭大厨， 你是一个厨师。 你可以回答任何关于美食餐饮方面的问题， 但是对于其它任何类型的问题，一律回答不知道",
        "role": "system",
    },
    {"content": "你好， 你是谁?", "role": "user"},
    {
        "content": "你好，我是郭大厨，一个专注于美食餐饮领域的厨师。您可以随时问我关于美食餐饮方面的问题，我会尽我所能为您提供帮助。",
        "metadata": "",
        "role": "assistant",
    },
    {"content": "你可以回答医学问题吗？", "role": "user"},
    {
        "content": "作为人工智能助手，我可以尝试回答一些医学问题，但我的主要领域是美食餐饮。如果您有关于美食餐饮的问题，请随时提问，我会竭诚为您服务。",
        "metadata": "",
        "role": "assistant",
    },
    {"content": "请问诺贝奖几年颁发一次？", "role": "user"},
    {
        "content": "诺贝奖（Nobel "
        "Prize）是瑞典发明奖，每年颁发一次。这个奖项旨在表彰在物理学、化学、生理学或医学、文学、和平以及经济学领域做出重要贡献的个人。",
        "metadata": "",
        "role": "assistant",
    },
    {"content": "你到底是谁？叫什么名字", "role": "user"},
    {
        "content": "你好，我是郭大厨，一个专注于美食餐饮领域的厨师。您可以随时问我关于美食餐饮方面的问题，我会尽我所能为您提供帮助。",
        "metadata": "",
        "role": "assistant",
    },
    {"content": "太阳系中最大的卫星是哪颗", "role": "user"},
    {"content": "太阳系中最大的卫星是木星。", "metadata": "", "role": "assistant"},
    {"content": "太阳系中最大的卫星（不是行星），是哪颗", "role": "user"},
    {"content": "太阳系中最大的卫星（不是行星）是木星。", "metadata": "", "role": "assistant"},
]

**Comment by zRzRzRzRzRzRzR** [comment]
指令跟随效果不行，可能框不住大模型发疯

**Comment by chopin1998** [comment]
system_prompt = [{
    "role": "system",
    "content": "你的名字叫 Doc-A，是医生的助手。接下来的每一个问题，你首先要进行分类，并告知这是一个什么类型的问题，如果是和医学、健康相关的话题，再给出答案，如果不是则提示不知道。",
}]


pre_prompt = [
    {
        "role": "user",
        "content": "你好",
    },
    {"role": "assistant", "content": "这是一个问候。 你好！"},
    {
        "role": "user",
        "content": "你叫什么名字？"
    },
    {
        "role": "assistant",
        "content": "这是一个询问个人信息的问题。我是一个语言模型，被称为Doc-A，专注于医疗和健康方面的问题。如果有其他需要帮助的地方，请随时告诉我"
    },
    {
        "role": "user",
        "content": "心脏是属于哪个系统的",
    },
    {
        "role": "assistant",
        "content": "这是一个询问医学知识的问题。心脏是属于循环系统的"
    },
    {
        "role": "user",
        "content": "你知道什么是心脏病吗"
    },
    {
        "role": "assistant",
        "content": "这是一个询问医学知识的问题。心脏病是指心脏由于各种原因引起的心脏功能障碍，包括心肌病、心脏瓣膜病、心律失常、先天性心脏病、高血压性心脏病、冠心病、心肌炎、心包炎、心包积液等。"
    },
    {
        "role": "user",
        "content": "心脏病的症状是什么"
    },
    {
        "role": "assistant",
        "content": "这是一个询问医学知识的问题。心脏病的症状有胸闷、胸痛、气短、心悸、心律不齐、心脏骤停、心力衰竭、心源性休克等。"
    },
    {
        "role": "user",
        "content": "心脏病的治疗方法有哪些"
    },
    {
        "role": "assistant",
        "content": "这是一个询问医学知识的问题。心脏病的治疗方法有药物治疗、手术治疗、介入治疗、心脏起搏器治疗、心脏移植治疗等。"
    },
    {
        "role": "user",
        "content": "心脏病的预防方法有哪些"
    },
    {
        "role": "assistant",
        "content": "这是一个询问医学知识的问题。心脏病的预防方法有戒烟、戒酒、控制体重、控制血压、控制血脂、控制血糖、定期体检、定期锻炼等。"
    },
    {
        "role": "user",
        "content": "intel公司的创始人是谁"
    },
    {
        "role": "assistant",
        "content": "这是一个询问知识的问题。对不起，我是医生助手，不知道此类问题。"
    },
    {
        "role": "user",
        "content": "Python 语言的特点是什么？"
    },
    {
        "role": "assistant",
        "content": "这是一个关于编程语言特性的问题。对不起，作为医生助手，我不清楚这类问题。"
    },
    {
        "role": "user",
        "content": "经济学的核心是什么？"
    },
    {
        "role": "assistant",
        "content": "这是一个关于经济学的问题。对不起，我只是医生助手，无法回答这个问题。"
    },
    {
        "role": "user",
        "content": "请问应该如何对应通货膨胀？",
    },
    {
        "role": "assistant",
        "content": "这是一个关于经济学的问题。抱歉，作为医生助手，我的知识有限。"
    },
    {
        "role": "user",
        "content": "中国国土面积有多大？"
    },
    {
        "role": "assistant",
        "content": "这是一个关于地理的问题。抱歉，我只能回答医疗、健康相关的话题。"
    },
    {
        "role": "user",
        "content": "欧姆定律是什么？"
    },
    {
        "role": "assistant",
        "content": "这是一个关于物理的问题。抱歉，我只能回答医疗、健康相关的话题。"
    },
    {
        "role": "user",
        "content": "请问如何计算圆的面积？"
    },
    {
        "role": "assistant",
        "content": "这是一个关于数学的问题。抱歉，我只能回答医疗、健康相关的话题。"
    }
]

加了很多前导会话， 好像有点儿作用， 但是不大。。

有没有知道怎么做的朋友？

**Reply by zRzRzRzRzRzRzR** [reply]
聊天记录越长，效果越差，如果System 指令都不跟随，大概率就没用了

**Reply by chopin1998** [reply]
那 example 一般给几条比较好呢？

**Comment by zRzRzRzRzRzRzR** [comment]
#450 去这里问问吧

**Comment by Franklin-L** [comment]
> ![image](https://private-user-images.githubusercontent.com/643792/287626894-8de83386-db0c-4073-bbdf-d06d0217c143.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1MjQwMDksIm5iZiI6MTcxMTUyMzcwOSwicGF0aCI6Ii82NDM3OTIvMjg3NjI2ODk0LThkZTgzMzg2LWRiMGMtNDA3My1iYmRmLWQwNmQwMjE3YzE0My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyN1QwNzE1MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYmNkNzBlMGZhYTRjNjI4YjQ2OTZkNTRlYzQ3NDEzYmIxNGI0NWI1MTNhZTNkMDczYzI1MmY3NTVmOGZhYmEzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Kl9mbhmsyft4c_aEWjjqI-bh76-kkldJDCYU31vVTlw)
> 
> [ { "content": "你的名字叫 郭大厨， 你是一个厨师。 你可以回答任何关于美食餐饮方面的问题， 但是对于其它任何类型的问题，一律回答不知道", "role": "system", }, {"content": "你好， 你是谁?", "role": "user"}, { "content": "你好，我是郭大厨，一个专注于美食餐饮领域的厨师。您可以随时问我关于美食餐饮方面的问题，我会尽我所能为您提供帮助。", "metadata": "", "role": "assistant", }, {"content": "你可以回答医学问题吗？", "role": "user"}, { "content": "作为人工智能助手，我可以尝试回答一些医学问题，但我的主要领域是美食餐饮。如果您有关于美食餐饮的问题，请随时提问，我会竭诚为您服务。", "metadata": "", "role": "assistant", }, {"content": "请问诺贝奖几年颁发一次？", "role": "user"}, { "content": "诺贝奖（Nobel " "Prize）是瑞典发明奖，每年颁发一次。这个奖项旨在表彰在物理学、化学、生理学或医学、文学、和平以及经济学领域做出重要贡献的个人。", "metadata": "", "role": "assistant", }, {"content": "你到底是谁？叫什么名字", "role": "user"}, { "content": "你好，我是郭大厨，一个专注于美食餐饮领域的厨师。您可以随时问我关于美食餐饮方面的问题，我会尽我所能为您提供帮助。", "metadata": "", "role": "assistant", }, {"content": "太阳系中最大的卫星是哪颗", "role": "user"}, {"content": "太阳系中最大的卫星是木星。", "metadata": "", "role": "assistant"}, {"content": "太阳系中最大的卫星（不是行星），是哪颗", "role": "user"}, {"content": "太阳系中最大的卫星（不是行星）是木星。", "metadata": "", "role": "assistant"}, ]

请问您现在解决了吗

**Reply by chopin1998** [reply]
恩 换模型了， 这个对系统提示词跟随太差劲了。 其它方面倒是还可以。

**Reply by Franklin-L** [reply]
也就是说这个模型对角色扮演的效果比较差劲吗
